webrender/src/prim_store.rs

```diff
@@ -580,6 +566,74 @@ impl PrimitiveStore {
         PrimitiveIndex(prim_index)
     }
 
+    pub fn resolve_primitives(&mut self, resource_cache: &ResourceCache) {
+        for prim_index in self.prims_to_resolve.drain(..) {
+            let metadata = &mut self.cpu_metadata[prim_index.0];
+
+            if let Some(mask_key) = metadata.mask_image {
+                let tex_cache = resource_cache.get_image(mask_key, ImageRendering::Auto);
+                metadata.mask_texture_id = tex_cache.texture_id;
+                if let Some(address) = metadata.clip_index {
+                    let clip = self.gpu_data32.get_slice_mut(address, 6);
+                    let old = clip[5].data; //TODO: avoid retaining the screen rectangle
+                    clip[5] = GpuBlock32::from(ImageMaskInfo {
+                        uv_rect: Rect::new(tex_cache.uv0,
+                                           Size2D::new(tex_cache.uv1.x - tex_cache.uv0.x,
+                                                       tex_cache.uv1.y - tex_cache.uv0.x)),
+                        local_rect: Rect::new(Point2D::new(old[4], old[5]),
+                                              Size2D::new(old[6], old[7])),
+                    });
+                }
+            }
+
+            match metadata.prim_kind {
+                PrimitiveKind::Rectangle |
+                PrimitiveKind::Border |
+                PrimitiveKind::BoxShadow |
+                PrimitiveKind::Gradient => {}
+                PrimitiveKind::TextRun => {
+                    let text = &mut self.cpu_text_runs[metadata.cpu_prim_index.0];
+
+                    let dest_glyphs = self.gpu_data32.get_slice_mut(metadata.gpu_data_address,
+                                                                    text.glyph_range.length);
+
+                    let texture_id = resource_cache.get_glyphs(text.font_key,
+                                                               text.font_size,
+                                                               text.blur_radius,
+                                                               &text.glyph_indices, |index, uv0, uv1| {
+                        let dest_glyph = &mut dest_glyphs[index];
+                        let dest: &mut GlyphPrimitive = unsafe {
+                            mem::transmute(dest_glyph)
```

Comment:
The existing data is a GpuBlock32 - I need to access it as the original type, in order to modify some fields of it. Is there a better way to express this (the From trait is used to create the GpuBlock32 initially)?